🙅 🦮 Remove device from regularizer

cthoyt opened the pull request:
This PR removes the device from the regularizer and extracts some functional code from the Lp and PowerSum normalizers

cthoyt committed:
Refactor regularizer class and test cases

cthoyt committed:
Update model test cases

Trigger CI

cthoyt committed:
Update test_regularizers.py

Trigger CI

cthoyt committed:
Reorganize testing assets

Trigger CI

cthoyt committed:
Remove device from Regularizer

cthoyt committed:
Merge branch 'master' into update-regularizers

cthoyt committed:
Update test_regularizers.py

Trigger CI

mberr commented on src/pykeen/regularizers.py:
```diff
@@ -46,20 +46,17 @@ class Regularizer(nn.Module, ABC):
         self.updated = False
         self.reset()
 
     def to(self, *args, **kwargs) -> Regularizer:  # noqa: D102
```
We do not need this method anymore - it was a dirty hack to manually move the regularization term along the module. Since we now use a buffer, pytorch knows what to do with it 🙂

cthoyt replied:
the to() method? okay lets nix it

mberr replied:
👍

mberr commented on src/pykeen/regularizers.py:
```diff
@@ -70,7 +67,7 @@ def get_normalized_name(cls) -> str:
     def reset(self) -> None:
         """Reset the regularization term to zero."""
-        self.regularization_term = torch.zeros(1, dtype=torch.float, device=self.device)
+        self.regularization_term = torch.zeros(1, dtype=torch.float)
```
Here, I am not sure whether it still stays a buffer, it we assign a new tensor to it. 

An option would be the fill the tensor in-place with zeros and detach?
```python
self.regularization_term.detach_().zeros_()
```

we should check (via the training unittest), whether this breaks gradient computation.

cthoyt replied:
This snipped didn't work, so i switched to 
```python
torch.zeros(size=self.regularization_term.shape, out=self.regularization_term)
```

which gave:
```python-traceback
Error
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/local/Cellar/python@3.8/3.8.6_2/Frameworks/Python.framework/Versions/3.8/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/Users/cthoyt/dev/pykeen/tests/cases.py", line 626, in test_model
    model.post_parameter_update()
  File "/Users/cthoyt/dev/pykeen/src/pykeen/models/base.py", line 850, in post_parameter_update
    super().post_parameter_update()
  File "/Users/cthoyt/dev/pykeen/src/pykeen/models/base.py", line 598, in post_parameter_update
    self.regularizer.reset()
  File "/Users/cthoyt/dev/pykeen/src/pykeen/regularizers.py", line 66, in reset
    torch.zeros(size=self.regularization_term.shape, out=self.regularization_term)
RuntimeError: cannot resize variables that require grad
```

cthoyt replied:
`self.regularization_term.detach_().zero_()` works!

cthoyt committed:
Remove to()

cthoyt committed:
Prepare for more elegant reset

The update in `reset()` to do `self.regularization_term.detach_().zeros_()` did not work, but this will allow us to try other magicks

cthoyt committed:
Fix test

Trigger CI

cthoyt committed:
Update zeroing

cthoyt committed:
Factor out norm functions

cthoyt committed:
Update regularizers.py

Trigger CI

cthoyt committed:
Update norm.py

Trigger CI

cthoyt committed:
Merge branch 'master' into update-regularizers

cthoyt committed:
Update norm.py

Trigger CI

cthoyt committed:
Fix test

Trigger CI

PyKEEN-bot committed:
Trigger CI

mberr committed:
Fix return type annotation

mberr committed:
Fox regularization_term not being a buffer

mberr committed:
Merge remote-tracking branch 'origin/update-regularizers' into update-regularizers

mberr committed:
simplify assignment

mberr committed:
Move regularizer test instance to device

mberr committed:
Fix random number generator being a cpu-only generator

mberr committed:
Fix tensors' and instance's device for TransH tests

mberr committed:
Fix regularizer's device for tests for updating only once

mberr commented on tests/cases.py:
```diff
@@ -642,8 +640,8 @@ def test_reset(self) -> None:
         # Generate random tensors
-        a = torch.rand(self.batch_size, 10, device=self.device, generator=self.generator)
-        b = torch.rand(self.batch_size, 20, device=self.device, generator=self.generator)
+        a = torch.rand(self.batch_size, 10, generator=self.generator).to(device=self.device)
```
@cthoyt this pattern occurs frequently for tests - maybe we should create a new helper method for this?

mberr replied:
some background: since we explcitly initialize the random number generator and save it to `self.generator` before we know about the device, the generator is a cpu-only generator. Sadly, we cannot simply move it to the device, but rather have to create the tensor on cpu and then move them to device.

cthoyt replied:
I think I looked into this before - it looks like the generator doesn't have support for device choice :/

cthoyt replied:
Done in 0d38b560. There are other functions in `pykeen.utils` that are only used in testing, we could also consider moving those over in a future PR

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Make testing utility function

Trigger CI
